File: 2501-Longest Square Streak in an Array.py
class Solution(object):
    def longestSquareStreak(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        seen = dict()
        nums.sort(reverse = True)
        res = -1
        
        for num in nums:
            if num*num in seen:
                seen[num] = seen[num*num] + 1
                res = max(res, seen[num])
            else:
                seen[num] = 1
        
        return res

File: C++/2501-Longest Square Streak in an Array.cpp
class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        sort(begin(nums), end(nums));
        unordered_map<int, int> hmap;
        int res = -1;
        for (int num:nums) {
            int half = sqrt(num);
            if (half*half == num && hmap.contains(half)) {
                hmap[num] = hmap[half]+1;
                res = max(res, hmap[num]);
            } else {
                hmap[num] = 1;
            }
        }

        return res;
    }
};